distributed/tests/test_client.py

```diff
@@ -6119,14 +6123,66 @@ async def test_wait_for_workers(c, s, a, b):
     start = time()
     await future
     assert time() < start + 1
-    await w.close()
 
     with pytest.raises(TimeoutError) as info:
         await c.wait_for_workers(n_workers=10, timeout="1 ms")
 
-    assert "2/10" in str(info.value).replace(" ", "")
+    assert "3/10" in str(info.value).replace(" ", "")
+    assert "1 ms" in str(info.value)
+
+    future = asyncio.ensure_future(c.wait_for_workers(n_workers=2))
+    await asyncio.sleep(0.22)  # 2 chances
+    assert future.done()
+    await w.close()
+
+
+@gen_cluster(client=True)
+async def test_wait_for_workers_max(c, s, a, b):
+    future = asyncio.ensure_future(c.wait_for_workers(n_workers=3, mode="at most"))
+    await asyncio.sleep(0.22)  # 2 chances
+    assert future.done()
+
+    w = await Worker(s.address)
+    start = time()
+    await future
+    assert time() < start + 1
+
+    with pytest.raises(TimeoutError) as info:
+        await c.wait_for_workers(n_workers=1, timeout="1 ms", mode="at most")
+
+    assert "3/1" in str(info.value).replace(" ", "")
+    assert "1 ms" in str(info.value)
+
+
+@gen_cluster(client=True)
+async def test_wait_for_workers_exactly(c, s, a, b):
```

Comment:
Similarly, I would have expected this to start a `wait_for_workers` task, ensure it's not done, add/remove workers to the correct number, and then ensure it is done. This doesn't seem to be testing behavior of workers joining/leaving while the `wait_for_workers` is happening.